Issues running Teku with OpenEthereum

<!-- Have you done the following? -->
<!--   * read the Code of Conduct? By filing an Issue, you are expected to -->  
<!--     comply with it, including treating everyone with respect: -->
<!--     https://github.com/PegasysEng/teku/blob/master/CODE-OF-CONDUCT.md -->
<!--   * Reproduced the issue in the latest version of the software -->
<!--   * Read the debugging wiki: https://github.com/PegasysEng/teku/wiki/debugging -->
<!--   * Duplicate Issue check:  https://github.com/search?q=+is%3Aissue+repo%3APegasysEng/Teku -->
<!-- Note:  Not all sections will apply to all issue types. -->

### Description
We've got a user reporting the following stack trace when running Teku with [OpenEthereum](https://openethereum.wiki/Getting-Synced/).  It appears that the eth1 node is missing historical blocks, which is causing teku to throw a NPE:
```
16:17:20.699 INFO  - Eth1 tracker successfully caught up to chain head
16:17:20.707 FATAL - Exiting due to fatal error in Eth1DepositManager
java.util.concurrent.CompletionException: tech.pegasys.teku.service.serviceutils.FatalServiceFailureException: java.util.concurrent.CompletionException: java.lang.NullPointerException
        at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314) ~[?:?]
        at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319) ~[?:?]
        at java.util.concurrent.CompletableFuture.uniExceptionally(CompletableFuture.java:990) ~[?:?]
        at java.util.concurrent.CompletableFuture$UniExceptionally.tryFire(CompletableFuture.java:970) ~[?:?]
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) ~[?:?]
        at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2137) ~[?:?]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$propagateResult$3(SafeFuture.java:152) ~[teku-infrastructure-async-20.11.1.jar:20.11.1]
        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859) ~[?:?]
        at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837) ~[?:?]
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) ~[?:?]
        at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2137) ~[?:?]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$propagateResult$3(SafeFuture.java:152) ~[teku-infrastructure-async-20.11.1.jar:20.11.1]
        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859) ~[?:?]
        at java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:883) ~[?:?]
        at java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2315) ~[?:?]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.whenComplete(SafeFuture.java:503) ~[teku-infrastructure-async-20.11.1.jar:20.11.1]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.whenComplete(SafeFuture.java:28) ~[teku-infrastructure-async-20.11.1.jar:20.11.1]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.propagateResult(SafeFuture.java:147) ~[teku-infrastructure-async-20.11.1.jar:20.11.1]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$exceptionallyCompose$25(SafeFuture.java:347) ~[teku-infrastructure-async-20.11.1.jar:20.11.1]
        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859) ~[?:?]
        at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837) ~[?:?]
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) ~[?:?]
        at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2137) ~[?:?]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$propagateResult$3(SafeFuture.java:152) ~[teku-infrastructure-async-20.11.1.jar:20.11.1]
        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859) ~[?:?]
        at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837) ~[?:?]
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) ~[?:?]
        at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2137) ~[?:?]
        at org.web3j.utils.Async.lambda$run$1(Async.java:38) ~[core-4.6.2.jar:?]
        at java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1800) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630) [?:?]
        at java.lang.Thread.run(Thread.java:832) [?:?]
Caused by: tech.pegasys.teku.service.serviceutils.FatalServiceFailureException: java.util.concurrent.CompletionException: java.lang.NullPointerException
        at tech.pegasys.teku.pow.Eth1DepositManager.lambda$start$3(Eth1DepositManager.java:74) ~[teku-pow-20.11.1.jar:20.11.1]
        at java.util.concurrent.CompletableFuture.uniExceptionally(CompletableFuture.java:986) ~[?:?]
        ... 30 more
Caused by: java.util.concurrent.CompletionException: java.lang.NullPointerException
        at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314) ~[?:?]
        at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319) ~[?:?]
        at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1155) ~[?:?]
        ... 29 more
Caused by: java.lang.NullPointerException
        at tech.pegasys.teku.pow.MinimumGenesisTimeBlockFinder.compareBlockTimestampToMinGenesisTime(MinimumGenesisTimeBlockFinder.java:87) ~[teku-pow-20.11.1.jar:20.11.1]
        at tech.pegasys.teku.pow.MinimumGenesisTimeBlockFinder.lambda$binarySearchLoop$0(MinimumGenesisTimeBlockFinder.java:60) ~[teku-pow-20.11.1.jar:20.11.1]
        at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1146) ~[?:?]
        ... 29 more
Teku is shutting down
16:17:20.727 INFO  - Stopped ServerConnector@52af750b{HTTP/1.1, (http/1.1)}{0.0.0.0:5051}
16:17:20.730 INFO  - Stopping Javalin ...
16:17:20.730 INFO  - Javalin has stopped
16:17:20.732 WARN  - Error shutting down connection manager
```

We should try to make teku's eth1 fetching logic more forgiving, if possible.

--- a/pow/src/main/java/tech/pegasys/teku/pow/Eth1DepositManager.java
+++ b/pow/src/main/java/tech/pegasys/teku/pow/Eth1DepositManager.java
@@ -18,6 +18,7 @@ import static tech.pegasys.teku.pow.MinimumGenesisTimeBlockFinder.notifyMinGenes
 
 import com.google.common.base.Preconditions;
 import java.math.BigInteger;
+import java.util.Optional;
 import java.util.concurrent.TimeUnit;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
@@ -40,6 +41,7 @@ public class Eth1DepositManager {
   private final Eth1DepositStorageChannel eth1DepositStorageChannel;
   private final DepositProcessingController depositProcessingController;
   private final MinimumGenesisTimeBlockFinder minimumGenesisTimeBlockFinder;
+  private final Optional<UInt64> depositContractDeployBlock;
 
   public Eth1DepositManager(
       final Eth1Provider eth1Provider,
@@ -47,13 +49,15 @@ public class Eth1DepositManager {
       final ValidatingEth1EventsPublisher eth1EventsPublisher,
       final Eth1DepositStorageChannel eth1DepositStorageChannel,
       final DepositProcessingController depositProcessingController,
-      final MinimumGenesisTimeBlockFinder minimumGenesisTimeBlockFinder) {
+      final MinimumGenesisTimeBlockFinder minimumGenesisTimeBlockFinder,
+      final Optional<UInt64> depositContractDeployBlock) {
     this.eth1Provider = eth1Provider;
     this.asyncRunner = asyncRunner;
     this.eth1EventsPublisher = eth1EventsPublisher;
     this.eth1DepositStorageChannel = eth1DepositStorageChannel;
     this.depositProcessingController = depositProcessingController;
     this.minimumGenesisTimeBlockFinder = minimumGenesisTimeBlockFinder;
+    this.depositContractDeployBlock = depositContractDeployBlock;
   }
 
   public void start() {
@@ -85,7 +89,7 @@ public class Eth1DepositManager {
     Preconditions.checkArgument(headBlock != null, "eth1 headBlock should be defined");
     Preconditions.checkArgument(
         replayDepositsResult != null, "eth1 replayDepositsResult should be defined");
-    BigInteger startBlockNumber = replayDepositsResult.getFirstUnprocessedBlockNumber();
+    BigInteger startBlockNumber = getFirstUnprocessedBlockNumber(replayDepositsResult);
     if (headBlock.getNumber().compareTo(startBlockNumber) >= 0) {
       if (isBlockAfterMinGenesis(headBlock)) {
         return headAfterMinGenesisMode(headBlock, replayDepositsResult);
@@ -98,7 +102,10 @@ public class Eth1DepositManager {
       depositProcessingController.startSubscription(startBlockNumber);
     } else {
       // preGenesisSubscription starts processing from the next block
-      preGenesisSubscription(replayDepositsResult.getLastProcessedBlockNumber());
+      final BigInteger depositContractStart =
+          depositContractDeployBlock.orElse(UInt64.ZERO).minusMinZero(1).bigIntegerValue();
+      preGenesisSubscription(
+          replayDepositsResult.getLastProcessedBlockNumber().max(depositContractStart));
     }
     return SafeFuture.COMPLETE;
   }
@@ -130,7 +137,7 @@ public class Eth1DepositManager {
 
     if (replayDepositsResult.isPastMinGenesisBlock()) {
       depositProcessingController.startSubscription(
-          replayDepositsResult.getFirstUnprocessedBlockNumber());
+          getFirstUnprocessedBlockNumber(replayDepositsResult));
       return SafeFuture.COMPLETE;
     }
 
@@ -151,7 +158,7 @@ public class Eth1DepositManager {
       final EthBlock.Block minGenesisTimeBlock, final ReplayDepositsResult replayDepositsResult) {
     return depositProcessingController
         .fetchDepositsInRange(
-            replayDepositsResult.getFirstUnprocessedBlockNumber(), minGenesisTimeBlock.getNumber())
+            getFirstUnprocessedBlockNumber(replayDepositsResult), minGenesisTimeBlock.getNumber())
         .thenApply(__ -> minGenesisTimeBlock);
   }
 
@@ -191,5 +198,12 @@ public class Eth1DepositManager {
     } else {
       return SafeFuture.completedFuture(number);
     }
+  }
+
+  private BigInteger getFirstUnprocessedBlockNumber(
+      final ReplayDepositsResult replayDepositsResult) {
+    return replayDepositsResult
+        .getFirstUnprocessedBlockNumber()
+        .max(depositContractDeployBlock.orElse(UInt64.ZERO).bigIntegerValue());
   }
 }

--- a/pow/src/main/java/tech/pegasys/teku/pow/MinimumGenesisTimeBlockFinder.java
+++ b/pow/src/main/java/tech/pegasys/teku/pow/MinimumGenesisTimeBlockFinder.java
@@ -17,6 +17,7 @@ import static tech.pegasys.teku.infrastructure.unsigned.UInt64.ONE;
 import static tech.pegasys.teku.infrastructure.unsigned.UInt64.ZERO;
 
 import java.math.BigInteger;
+import java.util.Optional;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 import org.apache.tuweni.bytes.Bytes32;
@@ -34,9 +35,12 @@ public class MinimumGenesisTimeBlockFinder {
   private static final UInt64 TWO = UInt64.valueOf(2);
 
   private final Eth1Provider eth1Provider;
+  private final Optional<UInt64> eth1DepositContractDeployBlock;
 
-  public MinimumGenesisTimeBlockFinder(Eth1Provider eth1Provider) {
+  public MinimumGenesisTimeBlockFinder(
+      Eth1Provider eth1Provider, Optional<UInt64> eth1DepositContractDeployBlock) {
     this.eth1Provider = eth1Provider;
+    this.eth1DepositContractDeployBlock = eth1DepositContractDeployBlock;
   }
 
   /**
@@ -47,7 +51,8 @@ public class MinimumGenesisTimeBlockFinder {
    */
   public SafeFuture<EthBlock.Block> findMinGenesisTimeBlockInHistory(
       final BigInteger headBlockNumber) {
-    return binarySearchLoop(new SearchContext(headBlockNumber));
+    return binarySearchLoop(
+        new SearchContext(eth1DepositContractDeployBlock.orElse(ZERO), headBlockNumber));
   }
 
   private SafeFuture<EthBlock.Block> binarySearchLoop(final SearchContext searchContext) {
@@ -107,10 +112,11 @@ public class MinimumGenesisTimeBlockFinder {
   }
 
   private static class SearchContext {
-    private UInt64 low = UInt64.ZERO;
+    private UInt64 low;
     private UInt64 high;
 
-    public SearchContext(final BigInteger headBlockNumber) {
+    public SearchContext(final UInt64 minSearchBlock, final BigInteger headBlockNumber) {
+      this.low = minSearchBlock;
       this.high = UInt64.valueOf(headBlockNumber);
     }
   }

--- a/pow/src/test/java/tech/pegasys/teku/pow/Eth1DepositManagerTest.java
+++ b/pow/src/test/java/tech/pegasys/teku/pow/Eth1DepositManagerTest.java
@@ -24,6 +24,7 @@ import static org.mockito.Mockito.when;
 import static tech.pegasys.teku.infrastructure.async.SafeFuture.COMPLETE;
 
 import java.math.BigInteger;
+import java.util.Optional;
 import org.apache.tuweni.bytes.Bytes32;
 import org.junit.jupiter.api.AfterAll;
 import org.junit.jupiter.api.BeforeAll;
@@ -74,7 +75,8 @@ class Eth1DepositManagerTest {
           eth1EventsChannel,
           eth1DepositStorageChannel,
           depositProcessingController,
-          minimumGenesisTimeBlockFinder);
+          minimumGenesisTimeBlockFinder,
+          Optional.empty());
 
   @BeforeAll
   static void setConstants() {

--- a/pow/src/test/java/tech/pegasys/teku/pow/MinimumGenesisTimeBlockFinderTest.java
+++ b/pow/src/test/java/tech/pegasys/teku/pow/MinimumGenesisTimeBlockFinderTest.java
@@ -14,10 +14,12 @@
 package tech.pegasys.teku.pow;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.argThat;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.when;
 
 import java.math.BigInteger;
+import java.util.Optional;
 import org.junit.jupiter.api.AfterAll;
 import org.junit.jupiter.api.BeforeAll;
 import org.junit.jupiter.api.Test;
@@ -29,10 +31,10 @@ import tech.pegasys.teku.util.config.Constants;
 
 public class MinimumGenesisTimeBlockFinderTest {
 
-  private Eth1Provider eth1Provider = mock(Eth1Provider.class);
+  private final Eth1Provider eth1Provider = mock(Eth1Provider.class);
 
   private MinimumGenesisTimeBlockFinder minimumGenesisTimeBlockFinder =
-      new MinimumGenesisTimeBlockFinder(eth1Provider);
+      new MinimumGenesisTimeBlockFinder(eth1Provider, Optional.empty());
 
   @BeforeAll
   static void setUp() {
@@ -51,6 +53,38 @@ public class MinimumGenesisTimeBlockFinderTest {
     final Block[] blocks = withBlockTimestamps(timestamps);
     final int minGenesisTime = 3500;
     final Block expectedMinGenesisTimeBlock = blocks[4];
+    assertMinGenesisBlock(blocks, minGenesisTime, expectedMinGenesisTimeBlock);
+  }
+
+  @Test
+  void shouldFindMinGenesisTimeWhenDeployBlockSpecified() {
+    final UInt64 deployBlock = UInt64.valueOf(2); // Block number
+    minimumGenesisTimeBlockFinder =
+        new MinimumGenesisTimeBlockFinder(eth1Provider, Optional.of(deployBlock));
+
+    final long[] timestamps = {0, 1000, 2000, 3000, 4000, 5000, 6000, 7000, 8000};
+    final Block[] blocks = withBlockTimestamps(timestamps);
+
+    when(eth1Provider.getEth1Block(argThat((UInt64 argument) -> argument.isLessThan(deployBlock))))
+        .thenReturn(SafeFuture.completedFuture(null));
+    final int minGenesisTime = 3500;
+    final Block expectedMinGenesisTimeBlock = blocks[4];
+    assertMinGenesisBlock(blocks, minGenesisTime, expectedMinGenesisTimeBlock);
+  }
+
+  @Test
+  void shouldFindMinGenesisTimeWhenDeployBlockSpecified_deployedAfterMinGenesisBlock() {
+    final UInt64 deployBlock = UInt64.valueOf(6); // Block number
+    minimumGenesisTimeBlockFinder =
+        new MinimumGenesisTimeBlockFinder(eth1Provider, Optional.of(deployBlock));
+
+    final long[] timestamps = {0, 1000, 2000, 3000, 4000, 5000, 6000, 7000, 8000};
+    final Block[] blocks = withBlockTimestamps(timestamps);
+
+    when(eth1Provider.getEth1Block(argThat((UInt64 argument) -> argument.isLessThan(deployBlock))))
+        .thenReturn(SafeFuture.completedFuture(null));
+    final int minGenesisTime = 3500;
+    final Block expectedMinGenesisTimeBlock = blocks[6];
     assertMinGenesisBlock(blocks, minGenesisTime, expectedMinGenesisTimeBlock);
   }
 

--- a/services/powchain/src/main/java/tech/pegasys/teku/services/powchain/PowchainService.java
+++ b/services/powchain/src/main/java/tech/pegasys/teku/services/powchain/PowchainService.java
@@ -16,6 +16,7 @@ package tech.pegasys.teku.services.powchain;
 import static tech.pegasys.teku.pow.api.Eth1DataCachePeriodCalculator.calculateEth1DataCacheDurationPriorToCurrentTime;
 import static tech.pegasys.teku.util.config.Constants.MAXIMUM_CONCURRENT_ETH1_REQUESTS;
 
+import java.util.Optional;
 import java.util.concurrent.TimeUnit;
 import okhttp3.OkHttpClient;
 import okhttp3.logging.HttpLoggingInterceptor;
@@ -25,6 +26,7 @@ import org.web3j.protocol.Web3j;
 import org.web3j.protocol.http.HttpService;
 import tech.pegasys.teku.infrastructure.async.AsyncRunner;
 import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
 import tech.pegasys.teku.pow.DepositContractAccessor;
 import tech.pegasys.teku.pow.DepositFetcher;
 import tech.pegasys.teku.pow.DepositProcessingController;
@@ -99,6 +101,8 @@ public class PowchainService extends Service {
             eth1BlockFetcher,
             headTracker);
 
+    final Optional<UInt64> eth1DepositContractDeployBlock =
+        tekuConfig.getEth1DepositContractDeployBlock();
     eth1DepositManager =
         new Eth1DepositManager(
             eth1Provider,
@@ -106,7 +110,8 @@ public class PowchainService extends Service {
             eth1EventsPublisher,
             eth1DepositStorageChannel,
             depositProcessingController,
-            new MinimumGenesisTimeBlockFinder(eth1Provider));
+            new MinimumGenesisTimeBlockFinder(eth1Provider, eth1DepositContractDeployBlock),
+            eth1DepositContractDeployBlock);
 
     chainIdValidator = new Eth1ChainIdValidator(eth1Provider, asyncRunner);
   }

--- a/storage/api/src/main/java/tech/pegasys/teku/storage/api/schema/ReplayDepositsResult.java
+++ b/storage/api/src/main/java/tech/pegasys/teku/storage/api/schema/ReplayDepositsResult.java
@@ -18,7 +18,7 @@ import java.util.Optional;
 
 public class ReplayDepositsResult {
   private static final BigInteger NEGATIVE_ONE = BigInteger.valueOf(-1);
-  private static ReplayDepositsResult EMPTY =
+  private static final ReplayDepositsResult EMPTY =
       new ReplayDepositsResult(NEGATIVE_ONE, Optional.empty(), false);
 
   private final BigInteger lastProcessedBlockNumber;

--- a/util/src/main/java/tech/pegasys/teku/util/config/GlobalConfiguration.java
+++ b/util/src/main/java/tech/pegasys/teku/util/config/GlobalConfiguration.java
@@ -14,9 +14,11 @@
 package tech.pegasys.teku.util.config;
 
 import java.util.List;
+import java.util.Optional;
 import org.apache.commons.lang3.StringUtils;
 import tech.pegasys.teku.infrastructure.logging.LoggingDestination;
 import tech.pegasys.teku.infrastructure.metrics.MetricsConfig;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
 
 /** @deprecated - Use TekuConfiguration where possible. Global application configuration. */
 @Deprecated
@@ -40,6 +42,7 @@ public class GlobalConfiguration implements MetricsConfig {
   private final Eth1Address eth1DepositContractAddress;
   private final String eth1Endpoint;
   private final boolean eth1DepositsFromStorageEnabled;
+  private final Optional<UInt64> eth1DepositContractDeployBlock;
 
   // Logging
   private final boolean logColorEnabled;
@@ -98,6 +101,7 @@ public class GlobalConfiguration implements MetricsConfig {
       final boolean interopEnabled,
       final Eth1Address eth1DepositContractAddress,
       final String eth1Endpoint,
+      final Optional<UInt64> eth1DepositContractDeployBlock,
       final boolean eth1DepositsFromStorageEnabled,
       final boolean logColorEnabled,
       final boolean logIncludeEventsEnabled,
@@ -139,6 +143,7 @@ public class GlobalConfiguration implements MetricsConfig {
     this.interopEnabled = interopEnabled;
     this.eth1DepositContractAddress = eth1DepositContractAddress;
     this.eth1Endpoint = eth1Endpoint;
+    this.eth1DepositContractDeployBlock = eth1DepositContractDeployBlock;
     this.eth1DepositsFromStorageEnabled = eth1DepositsFromStorageEnabled;
     this.logColorEnabled = logColorEnabled;
     this.logIncludeEventsEnabled = logIncludeEventsEnabled;
@@ -223,6 +228,10 @@ public class GlobalConfiguration implements MetricsConfig {
 
   public Eth1Address getEth1DepositContractAddress() {
     return eth1DepositContractAddress;
+  }
+
+  public Optional<UInt64> getEth1DepositContractDeployBlock() {
+    return eth1DepositContractDeployBlock;
   }
 
   public String getEth1Endpoint() {

--- a/util/src/main/java/tech/pegasys/teku/util/config/GlobalConfigurationBuilder.java
+++ b/util/src/main/java/tech/pegasys/teku/util/config/GlobalConfigurationBuilder.java
@@ -17,6 +17,7 @@ import java.util.List;
 import java.util.Optional;
 import java.util.function.Supplier;
 import tech.pegasys.teku.infrastructure.logging.LoggingDestination;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
 
 /**
  * @deprecated - Use TekuConfigurationBuilder where possible. Global application configuration
@@ -37,6 +38,7 @@ public class GlobalConfigurationBuilder {
   private boolean interopEnabled;
   private Eth1Address eth1DepositContractAddress;
   private String eth1Endpoint;
+  private Optional<UInt64> eth1DepositContractDeployBlock = Optional.empty();
   private boolean eth1DepositsFromStorageEnabled;
   private boolean logColorEnabled;
   private boolean logIncludeEventsEnabled;
@@ -129,6 +131,12 @@ public class GlobalConfigurationBuilder {
 
   public GlobalConfigurationBuilder setEth1Endpoint(final String eth1Endpoint) {
     this.eth1Endpoint = eth1Endpoint;
+    return this;
+  }
+
+  public GlobalConfigurationBuilder setEth1DepositContractDeployBlock(
+      final Optional<UInt64> eth1DepositContractDeployBlock) {
+    this.eth1DepositContractDeployBlock = eth1DepositContractDeployBlock;
     return this;
   }
 
@@ -297,6 +305,7 @@ public class GlobalConfigurationBuilder {
       eth1DepositContractAddress =
           getOrOptionalDefault(eth1DepositContractAddress, network::getEth1DepositContractAddress);
       eth1Endpoint = getOrOptionalDefault(eth1Endpoint, network::getEth1Endpoint);
+      eth1DepositContractDeployBlock = network.getEth1DepositContractDeployBlock();
     }
 
     if (eth1DepositContractAddress == null && eth1Endpoint != null) {
@@ -318,6 +327,7 @@ public class GlobalConfigurationBuilder {
         interopEnabled,
         eth1DepositContractAddress,
         eth1Endpoint,
+        eth1DepositContractDeployBlock,
         eth1DepositsFromStorageEnabled,
         logColorEnabled,
         logIncludeEventsEnabled,

--- a/util/src/main/java/tech/pegasys/teku/util/config/NetworkDefinition.java
+++ b/util/src/main/java/tech/pegasys/teku/util/config/NetworkDefinition.java
@@ -21,6 +21,7 @@ import java.util.ArrayList;
 import java.util.List;
 import java.util.Locale;
 import java.util.Optional;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
 
 public class NetworkDefinition {
   private static final ImmutableMap<String, NetworkDefinition> NETWORKS =
@@ -33,6 +34,7 @@ public class NetworkDefinition {
                   .initialStateFromClasspath("mainnet-genesis.ssz")
                   .startupTimeoutSeconds(120)
                   .eth1DepositContractAddress("0x00000000219ab540356cBB839Cbe05303d7705Fa")
+                  .eth1DepositContractDeployBlock(11052984)
                   .discoveryBootnodes(
                       // PegaSys Teku
                       "enr:-KG4QJRlj4pHagfNIm-Fsx9EVjW4rviuZYzle3tyddm2KAWMJBDGAhxfM2g-pDaaiwE8q19uvLSH4jyvWjypLMr3TIcEhGV0aDKQ9aX9QgAAAAD__________4JpZIJ2NIJpcIQDE8KdiXNlY3AyNTZrMaEDhpehBDbZjM_L9ek699Y7vhUJ-eAdMyQW_Fil522Y0fODdGNwgiMog3VkcIIjKA",
@@ -129,6 +131,7 @@ public class NetworkDefinition {
   private final List<String> discoveryBootnodes;
   private final Optional<Eth1Address> eth1DepositContractAddress;
   private final Optional<String> eth1Endpoint;
+  private final Optional<UInt64> eth1DepositContractDeployBlock;
 
   private NetworkDefinition(
       final String constants,
@@ -137,7 +140,8 @@ public class NetworkDefinition {
       final int startupTimeoutSeconds,
       final List<String> discoveryBootnodes,
       final Optional<Eth1Address> eth1DepositContractAddress,
-      final Optional<String> eth1Endpoint) {
+      final Optional<String> eth1Endpoint,
+      final Optional<UInt64> eth1DepositContractDeployBlock) {
     this.constants = constants;
     this.initialState = initialState;
     this.startupTargetPeerCount = startupTargetPeerCount;
@@ -145,6 +149,7 @@ public class NetworkDefinition {
     this.discoveryBootnodes = discoveryBootnodes;
     this.eth1DepositContractAddress = eth1DepositContractAddress;
     this.eth1Endpoint = eth1Endpoint;
+    this.eth1DepositContractDeployBlock = eth1DepositContractDeployBlock;
   }
 
   public static NetworkDefinition fromCliArg(final String arg) {
@@ -179,6 +184,10 @@ public class NetworkDefinition {
     return eth1DepositContractAddress;
   }
 
+  public Optional<UInt64> getEth1DepositContractDeployBlock() {
+    return eth1DepositContractDeployBlock;
+  }
+
   public Optional<String> getEth1Endpoint() {
     return eth1Endpoint;
   }
@@ -196,6 +205,7 @@ public class NetworkDefinition {
     private List<String> discoveryBootnodes = new ArrayList<>();
     private Optional<Eth1Address> eth1DepositContractAddress = Optional.empty();
     private Optional<String> eth1Endpoint = Optional.empty();
+    private Optional<UInt64> eth1DepositContractDeployBlock = Optional.empty();
 
     public Builder constants(final String constants) {
       this.constants = constants;
@@ -238,6 +248,12 @@ public class NetworkDefinition {
       return this;
     }
 
+    public Builder eth1DepositContractDeployBlock(final long eth1DepositContractDeployBlock) {
+      this.eth1DepositContractDeployBlock =
+          Optional.of(UInt64.valueOf(eth1DepositContractDeployBlock));
+      return this;
+    }
+
     public NetworkDefinition build() {
       checkNotNull(constants, "Missing constants");
       return new NetworkDefinition(
@@ -247,7 +263,8 @@ public class NetworkDefinition {
           startupTimeoutSeconds,
           discoveryBootnodes,
           eth1DepositContractAddress,
-          eth1Endpoint);
+          eth1Endpoint,
+          eth1DepositContractDeployBlock);
     }
   }
 }

